File: backend/service/finance.py
"""Cari hesaplamalar: FIFO borç kapatma, vade takibi ve aylık özet."""

import calendar
import datetime
from decimal import Decimal

from django.db.models import Q

from .models import Customer, Expense, FinanceSettings, LedgerEntry, ServiceReport

ZERO = Decimal("0")
MONTH_NAMES = [
    "Ocak", "Şubat", "Mart", "Nisan", "Mayıs", "Haziran",
    "Temmuz", "Ağustos", "Eylül", "Ekim", "Kasım", "Aralık",
]


def month_label(year, month):
    return f"{MONTH_NAMES[month - 1]} {year}"


def month_range(year, month):
    first = datetime.date(year, month, 1)
    last = datetime.date(year, month, calendar.monthrange(year, month)[1])
    return first, last


def customer_account(customer, entries=None, today=None, grace_days=0):
    """Bir müşterinin cari durumu.

    Tahsilat ve iadeler tarih sırasıyla en eski borçtan başlayarak düşülür;
    kalan açık borçlar vade/söz kontrolüne girer. `grace_days`, vade gününde
    hemen uyarı verilmemesi için beklenen tolerans süresidir.
    """
    today = today or datetime.date.today()
    if entries is None:
        entries = list(customer.ledger_entries.all())
    entries = sorted(entries, key=lambda e: (e.date, e.id))

    debits = [e for e in entries if e.type == "BORC"]
    credits = [e for e in entries if e.type != "BORC"]

    debit_total = sum((e.try_amount for e in debits), ZERO)
    credit_total = sum((e.try_amount for e in credits), ZERO)

    remaining = credit_total
    open_debts = []
    for debt in debits:
        paid = min(remaining, debt.try_amount)
        remaining -= paid
        open_amount = debt.try_amount - paid
        if open_amount > Decimal("0.005"):
            deadline = debt.promised_date or debt.due_date
            days_late = (today - deadline).days if deadline and deadline < today else 0
            is_overdue = days_late > grace_days
            open_debts.append({
                "entry": debt,
                "open_try": open_amount,
                "deadline": deadline,
                "is_overdue": is_overdue,
                "broken_promise": bool(is_overdue and debt.promised_date),
                "days_late": days_late,
            })

    overdue_total = sum((d["open_try"] for d in open_debts if d["is_overdue"]), ZERO)

    return {
        "customer": customer,
        "debit_try": debit_total,
        "credit_try": credit_total,
        "balance_try": debit_total - credit_total,
        "open_debts": open_debts,
        "overdue_try": overdue_total,
        "has_overdue": overdue_total > 0,
        "last_activity": entries[-1].date if entries else None,
    }


def _grace_days(value=None):
    if value is not None:
        return value
    return FinanceSettings.load().overdue_grace_days


def all_accounts(today=None, grace_days=None):
    grace_days = _grace_days(grace_days)
    customers = list(Customer.objects.all())
    entries = list(LedgerEntry.objects.select_related("customer"))
    grouped = {}
    for entry in entries:
        grouped.setdefault(entry.customer_id, []).append(entry)
    return [customer_account(c, grouped.get(c.id, []), today, grace_days) for c in customers]


def overdue_list(today=None, grace_days=None):
    """Tüm müşterilerdeki vadesi/sözü geçmiş açık alacaklar, en gecikmişten başlayarak."""
    rows = []
    for account in all_accounts(today, grace_days):
        for debt in account["open_debts"]:
            if debt["is_overdue"]:
                rows.append({"customer": account["customer"], **debt})
    return sorted(rows, key=lambda r: r["days_late"], reverse=True)


def monthly_summary(year, month):
    first, last = month_range(year, month)
    entries = LedgerEntry.objects.filter(date__gte=first, date__lte=last)
    expenses = Expense.objects.filter(date__gte=first, date__lte=last)

    income = [e for e in entries if e.type == "BORC"]
    collected = [e for e in entries if e.type == "TAHSILAT"]
    expense_list = list(expenses)

    income_total = sum((e.try_amount for e in income), ZERO)
    collected_total = sum((e.try_amount for e in collected), ZERO)
    expense_total = sum((e.try_amount for e in expense_list), ZERO)

    by_category = {}
    for expense in expense_list:
        by_category[expense.category] = by_category.get(expense.category, ZERO) + expense.try_amount

    by_currency = {}
    for entry in income:
        by_currency[entry.currency] = by_currency.get(entry.currency, ZERO) + entry.amount

    by_customer = {}
    for entry in income:
        row = by_customer.setdefault(entry.customer_id, {"income": ZERO, "collected": ZERO})
        row["income"] += entry.try_amount
    for entry in collected:
        row = by_customer.setdefault(entry.customer_id, {"income": ZERO, "collected": ZERO})
        row["collected"] += entry.try_amount

    names = dict(Customer.objects.values_list("id", "name"))
    category_labels = dict(Expense.CATEGORY_CHOICES)

    return {
        "year": year,
        "month": month,
        "label": month_label(year, month),
        "income_try": income_total,
        "collected_try": collected_total,
        "expense_try": expense_total,
        "billable_expense_try": sum((e.try_amount for e in expense_list if e.billable), ZERO),
        "net_try": income_total - expense_total,
        "cash_flow_try": collected_total - expense_total,
        "service_count": ServiceReport.objects.filter(
            service_date__gte=first, service_date__lte=last
        ).count(),
        "fuel_liters": sum((e.quantity for e in expense_list if e.category == "YAKIT"), ZERO),
        "expense_by_category": [
            {"category": key, "label": category_labels.get(key, key), "amount": value}
            for key, value in sorted(by_category.items(), key=lambda kv: kv[1], reverse=True)
        ],
        "income_by_currency": [
            {"currency": key, "amount": value} for key, value in by_currency.items()
        ],
        "by_customer": [
            {
                "customer_id": key,
                "customer": names.get(key, "-"),
                "income": value["income"],
                "collected": value["collected"],
            }
            for key, value in sorted(by_customer.items(), key=lambda kv: kv[1]["income"], reverse=True)
        ],
    }


def receivables_total():
    return sum(
        (a["balance_try"] for a in all_accounts() if a["balance_try"] > 0),
        ZERO,
    )

File: servis-app/app/src/main/java/com/technoral/servis/data/Finance.kt
package com.technoral.servis.data

import java.util.Calendar

/**
 * Cari hesaplamalar. Tahsilatlar, en eski borçtan başlayarak (FIFO) kapatılır;
 * böylece hangi alacağın hâlâ açık olduğu ve vadesinin geçip geçmediği çıkarılabilir.
 */

/**
 * Açık kalan bir borç kalemi.
 *
 * [graceDays] ayarlardan gelen tolerans: vade gününde hemen uyarı vermemek için
 * bu kadar gün beklenir.
 */
data class OpenDebt(
    val entry: LedgerEntry,
    val openTry: Double,
    val graceDays: Int = 0,
) {
    /** Açık tutarın kendi para birimindeki karşılığı. */
    val openAmount: Double get() = if (entry.rate > 0) openTry / entry.rate else openTry

    private val deadline: Long?
        get() = entry.promisedDate ?: entry.dueDate

    val isOverdue: Boolean
        get() = daysLate > graceDays

    /** Vade mi yoksa verilen söz mü geçti? */
    val brokenPromise: Boolean
        get() = isOverdue && entry.promisedDate != null

    val daysLate: Long
        get() {
            val due = deadline ?: return 0
            return ((todayStart() - due) / 86_400_000L).coerceAtLeast(0)
        }
}

data class CustomerAccount(
    val customerId: String,
    val debitTry: Double,
    val creditTry: Double,
    val openDebts: List<OpenDebt>,
    val lastActivity: Long?,
) {
    val balanceTry: Double get() = debitTry - creditTry
    val overdueTry: Double get() = openDebts.filter { it.isOverdue }.sumOf { it.openTry }
    val hasOverdue: Boolean get() = openDebts.any { it.isOverdue }
    val brokenPromises: List<OpenDebt> get() = openDebts.filter { it.brokenPromise }
}

data class MonthlySummary(
    val year: Int,
    val month: Int,
    val incomeTry: Double,
    val collectedTry: Double,
    val expenseTry: Double,
    val billableExpenseTry: Double,
    val serviceCount: Int,
    val fuelLiters: Double,
    val expenseByCategory: List<Pair<ExpenseCategory, Double>>,
    val incomeByCurrency: List<Pair<Currency, Double>>,
    val topCustomers: List<Triple<String, Double, Double>>, // müşteri kimliği, hakediş, tahsilat
) {
    /** Dönem kârı: hakediş - masraf. */
    val netTry: Double get() = incomeTry - expenseTry

    /** Kasa hareketi: tahsil edilen - harcanan. */
    val cashFlowTry: Double get() = collectedTry - expenseTry

    val label: String get() = "${monthNames[month]} $year"
}

val monthNames = listOf(
    "Ocak", "Şubat", "Mart", "Nisan", "Mayıs", "Haziran",
    "Temmuz", "Ağustos", "Eylül", "Ekim", "Kasım", "Aralık",
)

fun todayStart(): Long {
    val c = Calendar.getInstance()
    c.set(Calendar.HOUR_OF_DAY, 0)
    c.set(Calendar.MINUTE, 0)
    c.set(Calendar.SECOND, 0)
    c.set(Calendar.MILLISECOND, 0)
    return c.timeInMillis
}

fun monthRange(year: Int, month: Int): LongRange {
    val start = Calendar.getInstance().apply {
        set(year, month, 1, 0, 0, 0)
        set(Calendar.MILLISECOND, 0)
    }
    val end = (start.clone() as Calendar).apply { add(Calendar.MONTH, 1) }
    return start.timeInMillis until end.timeInMillis
}

fun currentYearMonth(): Pair<Int, Int> {
    val c = Calendar.getInstance()
    return c.get(Calendar.YEAR) to c.get(Calendar.MONTH)
}

object Finance {

    /**
     * Bir müşterinin cari durumunu çıkarır. Tahsilat ve iadeler tarih sırasıyla
     * en eski borçtan başlayarak düşülür; kalan açık borçlar vade kontrolüne girer.
     */
    fun accountOf(
        customerId: String,
        entries: List<LedgerEntry>,
        graceDays: Int = 0,
    ): CustomerAccount {
        val own = entries.filter { it.customerId == customerId }.sortedBy { it.date }
        val debits = own.filter { it.type == LedgerType.BORC }
        val credits = own.filter { it.type != LedgerType.BORC }

        val debitTry = debits.sumOf { it.tryAmount }
        val creditTry = credits.sumOf { it.tryAmount }

        var remaining = creditTry
        val openDebts = mutableListOf<OpenDebt>()
        debits.forEach { debt ->
            val paid = minOf(remaining, debt.tryAmount)
            remaining -= paid
            val open = debt.tryAmount - paid
            if (open > 0.005) openDebts.add(OpenDebt(debt, open, graceDays))
        }

        return CustomerAccount(
            customerId = customerId,
            debitTry = debitTry,
            creditTry = creditTry,
            openDebts = openDebts,
            lastActivity = own.maxOfOrNull { it.date },
        )
    }

    fun accounts(
        customers: List<Customer>,
        entries: List<LedgerEntry>,
        graceDays: Int = 0,
    ): List<CustomerAccount> = customers.map { accountOf(it.id, entries, graceDays) }

    /** Tüm müşterilerdeki vadesi geçmiş açık alacaklar, en gecikmişten başlayarak. */
    fun overdueDebts(
        customers: List<Customer>,
        entries: List<LedgerEntry>,
        graceDays: Int = 0,
    ): List<Pair<String, OpenDebt>> =
        customers.flatMap { customer ->
            accountOf(customer.id, entries, graceDays).openDebts
                .filter { it.isOverdue }
                .map { customer.id to it }
        }.sortedByDescending { it.second.daysLate }

    fun totalReceivableTry(customers: List<Customer>, entries: List<LedgerEntry>): Double =
        customers.sumOf { accountOf(it.id, entries).balanceTry.coerceAtLeast(0.0) }

    /** Yalnızca gecikmiş alacakların toplamı. */
    fun overdueTotalTry(
        customers: List<Customer>,
        entries: List<LedgerEntry>,
        graceDays: Int = 0,
    ): Double = overdueDebts(customers, entries, graceDays).sumOf { it.second.openTry }

    fun summary(
        year: Int,
        month: Int,
        entries: List<LedgerEntry>,
        expenses: List<Expense>,
        reports: List<ServiceReport>,
    ): MonthlySummary {
        val range = monthRange(year, month)
        val periodEntries = entries.filter { it.date in range }
        val periodExpenses = expenses.filter { it.date in range }

        val income = periodEntries.filter { it.type == LedgerType.BORC }
        val collected = periodEntries.filter { it.type == LedgerType.TAHSILAT }

        val byCustomer = mutableMapOf<String, Pair<Double, Double>>()
        income.forEach { entry ->
            val current = byCustomer[entry.customerId] ?: (0.0 to 0.0)
            byCustomer[entry.customerId] = (current.first + entry.tryAmount) to current.second
        }
        collected.forEach { entry ->
            val current = byCustomer[entry.customerId] ?: (0.0 to 0.0)
            byCustomer[entry.customerId] = current.first to (current.second + entry.tryAmount)
        }

        return MonthlySummary(
            year = year,
            month = month,
            incomeTry = income.sumOf { it.tryAmount },
            collectedTry = collected.sumOf { it.tryAmount },
            expenseTry = periodExpenses.sumOf { it.tryAmount },
            billableExpenseTry = periodExpenses.filter { it.billable }.sumOf { it.tryAmount },
            serviceCount = reports.count { it.serviceDate in range },
            fuelLiters = periodExpenses.filter { it.category == ExpenseCategory.YAKIT }.sumOf { it.quantity },
            expenseByCategory = periodExpenses
                .groupBy { it.category }
                .map { (category, list) -> category to list.sumOf { it.tryAmount } }
                .sortedByDescending { it.second },
            incomeByCurrency = income
                .groupBy { it.currency }
                .map { (currency, list) -> currency to list.sumOf { it.amount } }
                .sortedByDescending { it.first.ordinal },
            topCustomers = byCustomer.entries
                .sortedByDescending { it.value.first }
                .take(8)
                .map { Triple(it.key, it.value.first, it.value.second) },
        )
    }

    /** Son 12 ayın hakediş/masraf serisi — özet ekranındaki mini grafik için. */
    fun lastMonths(
        count: Int,
        entries: List<LedgerEntry>,
        expenses: List<Expense>,
        reports: List<ServiceReport>,
    ): List<MonthlySummary> {
        val c = Calendar.getInstance()
        return (0 until count).map { offset ->
            val cursor = (c.clone() as Calendar).apply { add(Calendar.MONTH, -offset) }
            summary(
                cursor.get(Calendar.YEAR), cursor.get(Calendar.MONTH),
                entries, expenses, reports,
            )
        }.reversed()
    }
}
